### DISCUSSION TITLE
Problems with hooks in typescript on version 7

### DISCUSSION THREAD

**Discussion by kevin-longmate** [webdriverio/webdriverio#6443] [Q&A]
Hi,

Part of the tests I'm running uses a number of typescript files to build the config file up - for example I've got _capabilities.ts_ and _browserstack.ts_ for the config specific to browserstack, etc. When I'm running tests I'm essentially running the code in browserstack.ts rather than wdio.conf.js.

With the update to version 7 the code in the hooks are giving me issues.
```
    before: function (capabilities, specs) {
...
    },
```
will complain about capabilities and specs saying _Parameter 'capabilities' implicitly has an 'any' type._
This is true for all hooks - before, after, beforeSuite, etc.

I have "noImplicitAny" as true in tsconfig.json which is why it's throwing up an error, but this is a recommended setting so I would expect it to have been ok.

Should I change these config files to Javascript, remove _noImplicitAny_, or are there future plans to make these hooks work with _noImplicitAny_ turned on and I should just wait it out?

**Comment by portyu9** [comment]
Hi @kevin-longmate 

You shouldn't switch these files to JavaScript or disable `noImplicitAny`.

The issue is that TypeScript loses the contextual WebdriverIO hook types when the hooks are defined inside an otherwise untyped configuration fragment. Give that fragment a WebdriverIO config type and the hook parameters will be inferred.

For example:

```ts
import type { Options } from '@wdio/types'

export const browserstackConfig: Partial<Options.Testrunner> = {
    before(capabilities, specs) {
        // capabilities and specs are typed
    },

    beforeSuite(suite) {
        // suite is typed
    }
}
```

Then merge that object into your main WDIO config as you already do.

With current WebdriverIO, TypeScript execution is handled through `tsx`, but that doesn't change the recommendation: keep `noImplicitAny` enabled and type your configuration objects. WebdriverIO also recommends running `tsc` separately for actual type checking.

I reviewed the complete discussion; there were no replies, screenshots, or other attachments.

Please mark this answer as accepted if it helped, thank you.
